File: tests/oracle/database_symbol_subclass_1204.cc
/*
 * Locked Ghidra 12.0.4 oracle for DATABASE-SYMBOL-SUBCLASS-FIXTURE-0001
 * (MIGW1-DATABASE-0005 phase 2, second fixture).
 *
 * Exercises the Symbol-subclass construction surface and the entry/
 * comparator carriers Rust-ized in src/database.rs under the same
 * anchors:
 *
 *   FunctionSymbol::buildType        database.cc:514
 *   FunctionSymbol::FunctionSymbol   database.cc:534 / 545
 *   FunctionSymbol::getBytesConsumed  database.cc:508 override
 *   LabSymbol::buildType             database.cc:728
 *   LabSymbol::LabSymbol             database.cc:736 / 745
 *   ExternRefSymbol::buildNameType   database.cc:768
 *   ExternRefSymbol ctor             database.cc:789
 *   ExternRefSymbol::getRefAddr      database.hh:351
 *   UnionFacetSymbol ctor            database.cc:691 / hh:323
 *   UnionFacetSymbol::getFieldNumber database.hh:324
 *   Symbol::getBytesConsumed         database.cc:508
 *   Symbol::getMapEntryPosition      database.cc:301 (cc:309 quirk)
 *   SymbolEntry::getFirstUseAddress  database.cc:122
 *   SymbolEntry::printEntry          database.cc:166
 *   SymbolEntry::EntrySubsort        database.hh:107-134
 *   SymbolCompareName                database.hh:366
 *   DuplicateFunctionError          database.hh:435
 *   Scope::printBounds               database.hh:789
 *
 * All projections are id/name/size/message-level.  printEntry strings
 * are compared verbatim; the entry types are deliberately base types
 * (int4 / xunknown4) whose printRaw is the plain name on both sides —
 * the printRaw of code/union/struct/enum metatypes is outside this
 * fixture (a pre-existing cross-file deviation is reported separately).
 */

#include "bfd_arch.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "space.hh"
#include "type.hh"

#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::string;
using std::vector;

class FixtureArchitecture final : public BfdArchitecture {
public:
  FixtureArchitecture(const string &filename, const string &target,
                      std::ostream *estream)
    : BfdArchitecture(filename, target, estream) {}
};

class FixtureDatabase final : public Database {
public:
  FixtureDatabase(Architecture *g, bool idByName) : Database(g, idByName) {}
};

class FixtureSymbol final : public Symbol {
public:
  FixtureSymbol(Scope *sc, const string &nm, Datatype *ct)
    : Symbol(sc, nm, ct) {}
  void setDedup(uint4 v) { nameDedup = v; }
};

class FixtureScope final : public ScopeInternal {
public:
  FixtureScope(uint8 id, const string &name, Architecture *arch)
    : ScopeInternal(id, name, arch) {}
  FixtureSymbol *addRaw(const string &nm, Datatype *ct) {
    FixtureSymbol *sym = new FixtureSymbol(this, nm, ct);
    addSymbolInternal(sym);
    return sym;
  }
  // Expose the protected storage installer for partial-entry cases.
  SymbolEntry *addPiece(Symbol *sym, const Address &addr, int4 off, int4 sz,
                        const RangeList &uselim) {
    return addMapInternal(sym, 0, addr, off, sz, uselim);
  }
};

/// A ram-like processor space with deterministic attributes: 4-byte
/// addresses, word size 1, index 3, unassigned shortcut (' ').
AddrSpace *makeRam(void)
{
  return new AddrSpace((AddrSpaceManager *)0, (const Translate *)0,
                       IPTR_PROCESSOR, "ram", false, 4, 1, 3,
                       AddrSpace::hasphysical, -1, -1);
}

string hexoff(uintb v)
{
  std::ostringstream s;
  s << std::hex << v;
  return s.str();
}

void run(const string &specDirectory, const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  {
    std::ostringstream diagnostics;
    FixtureArchitecture arch(binary, "default", &diagnostics);
    DocumentStorage store;
    arch.init(store);
    if (arch.archid != "x86:LE:64:default:gcc")
      throw std::runtime_error("runtime architecture/compiler drifted: " + arch.archid);
    TypeFactory *types = arch.types;

    FixtureScope *global = new FixtureScope(100, "", &arch);
    FixtureDatabase db(&arch, false);
    db.attachScope(global, (Scope *)0);
    AddrSpace *ram = makeRam();

    // ---- FunctionSymbol (database.cc:514/534/545/508) ----
    {
      FunctionSymbol *f = new FunctionSymbol(global, "main", 2);
      std::cout << "case=fsym_buildtype|metatype=code|size=" << f->getType()->getSize()
                << "|namelock=" << ((f->getFlags() & Varnode::namelock) ? 1 : 0)
                << "|typelock=" << ((f->getFlags() & Varnode::typelock) ? 1 : 0)
                << "|name=" << f->getName() << '\n';
      std::cout << "case=fsym_bytes_consumed|" << f->getBytesConsumed() << '\n';
      std::cout << "case=fsym_type_size|" << f->getType()->getSize() << '\n';
      FunctionSymbol *fd = new FunctionSymbol(global, 3);
      std::cout << "case=fsym_decode_ctor|name_empty="
                << (fd->getName().size() == 0 ? 1 : 0)
                << "|consume=" << fd->getBytesConsumed() << '\n';
    }

    // ---- LabSymbol (database.cc:728/736/745) ----
    {
      LabSymbol *l = new LabSymbol(global, "loop");
      std::cout << "case=lsym_buildtype|typename=" << l->getType()->getName()
                << "|size=" << l->getType()->getSize()
                << "|name=" << l->getName() << '\n';
      LabSymbol *ld = new LabSymbol(global);
      std::cout << "case=lsym_decode_ctor|name_empty="
                << (ld->getName().size() == 0 ? 1 : 0)
                << "|typename=" << ld->getType()->getName() << '\n';
    }

    // ---- ExternRefSymbol (database.cc:768/789; hh:351) ----
    {
      Address ref(ram, 0x6000);
      ExternRefSymbol *x = new ExternRefSymbol(global, ref, "");
      std::cout << "case=exref_autoname|name=" << x->getName()
                << "|externref=" << ((x->getFlags() & Varnode::externref) ? 1 : 0)
                << "|typelock=" << ((x->getFlags() & Varnode::typelock) ? 1 : 0)
                << "|ptr_size=" << x->getType()->getSize()
                << "|ptr_meta=ptr" << '\n';
      Address ref2(ram, 0x6010);
      ExternRefSymbol *y = new ExternRefSymbol(global, ref2, "printf");
      std::cout << "case=exref_named|name=" << y->getName() << '\n';
      std::cout << "case=exref_ref_addr|" << hexoff(x->getRefAddr().getOffset()) << '\n';
    }

    // ---- UnionFacetSymbol (database.cc:691; hh:323/324) ----
    {
      Datatype *udt = types->getTypeUnion("fixture_union");
      UnionFacetSymbol *u = new UnionFacetSymbol(global, "f", udt, 2);
      std::cout << "case=ufacet_ctor|field=" << u->getFieldNumber()
                << "|category=" << u->getCategory()
                << "|typename=" << u->getType()->getName() << '\n';
      UnionFacetSymbol *ud = new UnionFacetSymbol(global);
      std::cout << "case=ufacet_decode|field=" << ud->getFieldNumber()
                << "|category=" << ud->getCategory() << '\n';
    }

    // ---- Symbol::getBytesConsumed (database.cc:508) ----
    {
      Datatype *int4 = types->getBase(4, TYPE_INT);
      FixtureSymbol *s = new FixtureSymbol(global, "s", int4);
      std::cout << "case=sym_bytes_consumed|" << s->getBytesConsumed() << '\n';
    }

    // ---- Symbol::getMapEntryPosition (database.cc:301-313) ----
    {
      Datatype *int4 = types->getBase(4, TYPE_INT);
      FixtureSymbol *m = global->addRaw("m", int4);
      Address a1(ram, 0x1000), a2(ram, 0x2000), a3(ram, 0x3000);
      RangeList empty;
      SymbolEntry *e1 = global->addMapPoint(m, a1, Address());
      SymbolEntry *e2 = global->addMapPoint(m, a2, Address());
      global->addMapPoint(m, a3, Address());
      // cc:309 quirk: the counter condition reads the SOUGHT entry's
      // size vs the type size — a whole-sized sought entry yields its
      // index among ALL entries; a partial sought entry always yields 0.
      SymbolEntry *piece = global->addPiece(m, Address(ram, 0x4000), 0, 2, empty);
      std::cout << "case=mapentry_whole_sought|" << m->getMapEntryPosition(e2) << '\n';
      std::cout << "case=mapentry_partial_sought|" << m->getMapEntryPosition(piece) << '\n';
      // An entry of a DIFFERENT symbol is absent from m's mapentry list.
      FixtureSymbol *other = global->addRaw("other", int4);
      SymbolEntry *foreign = global->addMapPoint(other, Address(ram, 0x8000), Address());
      std::cout << "case=mapentry_absent|" << m->getMapEntryPosition(foreign) << '\n';
      (void)e1;
    }

    // ---- SymbolEntry::getFirstUseAddress (database.cc:122) ----
    {
      Datatype *int4 = types->getBase(4, TYPE_INT);
      FixtureSymbol *m2 = new FixtureSymbol(global, "m2", int4);
      RangeList uselim;
      uselim.insertRange(ram, 0x4000, 0x4fff);
      SymbolEntry *e = global->addMapPoint(m2, Address(ram, 0x2000), Address());
      e->setUseLimit(uselim);
      std::cout << "case=firstuse_present|" << hexoff(e->getFirstUseAddress().getOffset()) << '\n';
      RangeList empty;
      SymbolEntry *e2 = global->addMapPoint(m2, Address(ram, 0x2100), Address());
      e2->setUseLimit(empty);
      std::cout << "case=firstuse_empty|" << (e2->getFirstUseAddress().isInvalid() ? 1 : 0) << '\n';
    }

    // ---- SymbolEntry::printEntry (database.cc:166-181) ----
    {
      std::ostringstream p1, p2, p3;
      Datatype *int4 = types->getBase(4, TYPE_INT);
      FixtureSymbol *m = global->addRaw("m", int4);
      SymbolEntry *e1 = global->addMapPoint(m, Address(ram, 0x1000), Address());
      e1->printEntry(p1);
      std::cout << "case=printentry_static_all|" << p1.str();

      FixtureSymbol *m2 = global->addRaw("m2", int4);
      RangeList uselim;
      uselim.insertRange(ram, 0x4000, 0x4fff);
      SymbolEntry *e2 = global->addMapPoint(m2, Address(ram, 0x2000), Address());
      e2->setUseLimit(uselim);
      e2->printEntry(p2);
      std::cout << "case=printentry_uselimit|" << p2.str();

      FixtureSymbol *d = global->addRaw("d", int4);
      RangeList empty;
      // The dynamic SymbolEntry ctor (database.hh:140) — dynamic entries
      // are never installed through addMapInternal (which dereferences the
      // entry address's space); printEntry only reads fields.
      SymbolEntry *e3 = new SymbolEntry(d, Varnode::mapped, 0x1234, 0, 4, empty);
      e3->printEntry(p3);
      std::cout << "case=printentry_dynamic|" << p3.str();
    }

    // ---- SymbolEntry::EntrySubsort (database.hh:107-134) ----
    {
      SymbolEntry::EntrySubsort earliest;
      SymbolEntry::EntrySubsort latest(true);
      std::cout << "case=subsort_earliest_lt_latest|" << (earliest < latest ? 1 : 0) << '\n';
      std::cout << "case=subsort_latest_gt_earliest|" << (latest < earliest ? 0 : 1) << '\n';
      Address sa(ram, 0x5000);
      Address sb(ram, 0x6000);
      SymbolEntry::EntrySubsort from_a(sa);
      SymbolEntry::EntrySubsort from_b(sb);
      std::cout << "case=subsort_addr_offset_order|" << (from_a < from_b ? 1 : 0) << '\n';
      std::cout << "case=subsort_addr_offset_rev|" << (from_b < from_a ? 0 : 1) << '\n';
      SymbolEntry::EntrySubsort from_a2(sa);
      std::cout << "case=subsort_addr_equal|" << (from_a < from_a2 ? 0 : 1) << '\n';
    }

    // ---- SymbolCompareName (database.hh:366) ----
    {
      Datatype *int4 = types->getBase(4, TYPE_INT);
      FixtureSymbol *s1 = new FixtureSymbol(global, "apple", int4);
      FixtureSymbol *s2 = new FixtureSymbol(global, "banana", int4);
      SymbolCompareName cmp;
      std::cout << "case=symcmp_name_order|" << (cmp(s1, s2) ? 1 : 0) << '\n';
      std::cout << "case=symcmp_name_rev|" << (cmp(s2, s1) ? 0 : 1) << '\n';
      FixtureSymbol *t1 = new FixtureSymbol(global, "same", int4);
      FixtureSymbol *t2 = new FixtureSymbol(global, "same", int4);
      t1->setDedup(1);
      t2->setDedup(2);
      std::cout << "case=symcmp_dedup_tiebreak|" << (cmp(t1, t2) ? 1 : 0) << '\n';
      std::cout << "case=symcmp_dedup_rev|" << (cmp(t2, t1) ? 0 : 1) << '\n';
    }

    // ---- DuplicateFunctionError (database.hh:435) ----
    {
      DuplicateFunctionError err(Address(ram, 0x5000), "foo");
      std::cout << "case=dupfn|msg=" << err.explain
                << "|addr=" << hexoff(err.address.getOffset())
                << "|name=" << err.functionName << '\n';
    }

    // ---- Scope::printBounds (database.hh:789) ----
    {
      FixtureScope *ns = new FixtureScope(101, "ns", &arch);
      db.attachScope(ns, global);
      RangeList rlist;
      rlist.insertRange(ram, 0x3000, 0x3fff);
      rlist.insertRange(ram, 0x1000, 0x1fff);
      db.setRange(ns, rlist);
      std::ostringstream pb;
      ns->printBounds(pb);
      std::cout << "case=scope_printbounds|" << pb.str();
    }

    delete ram;
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc, char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: database_symbol_subclass_1204 SPEC_ROOT BINARY" << std::endl;
    return 2;
  }
  try {
    run(argv[1], argv[2]);
    return 0;
  } catch (const LowlevelError &error) {
    std::cerr << "database_symbol_subclass_1204: LowlevelError: " << error.explain << std::endl;
    return 1;
  } catch (const std::exception &error) {
    std::cerr << "database_symbol_subclass_1204: " << error.what() << std::endl;
    return 1;
  }
}

File: tests/oracle/database_symbol_subclass_1204.rs
/*
 * Rugra comparand for DATABASE-SYMBOL-SUBCLASS-FIXTURE-0001
 * (MIGW1-DATABASE-0005 phase 2, second fixture), the bilateral twin of
 * tests/oracle/database_symbol_subclass_1204.cc.
 *
 * Exercises the same Symbol-subclass construction surface and the
 * entry/comparator carriers under the same anchors:
 *
 *   FunctionSymbol::buildType / ctors / getBytesConsumed
 *   LabSymbol::buildType / ctors
 *   ExternRefSymbol::buildNameType / ctor / getRefAddr
 *   UnionFacetSymbol ctors / getFieldNumber
 *   Symbol::getBytesConsumed / getMapEntryPosition (cc:309 quirk)
 *   SymbolEntry::getFirstUseAddress / printEntry
 *   SymbolEntry::EntrySubsort ordering
 *   SymbolCompareName / DuplicateFunctionError / Scope::printBounds
 *
 * All projections are id/name/size/message-level; printEntry strings are
 * compared verbatim.  The entry types are base types whose printRaw is
 * the plain name on both sides.  The TypeFactory uses the standalone
 * SLEIGH core table (CoreTypeFlavor::Standalone = sleigh_arch.cc:229),
 * mirroring the oracle's BfdArchitecture environment.
 *
 * Carrier adjudications mirrored here (recorded in docs/api/database.md):
 * - The C++ FunctionSymbol(Scope*,int4) decode ctor's observable state
 *   (empty name, consume size, code type, namelock|typelock) is reached
 *   through the named ctor with an empty name in the Rust value model —
 *   identical fields, construction-path difference only.
 * - The C++ Symbol carries its mapentry list; the Rust
 *   Symbol::get_map_entry_position takes the entry list as a parameter,
 *   so the fixture assembles the same insertion-ordered list.
 */

use rugra::address::{Address, Range, RangeList};
use rugra::database::{
    symbol_flags, DuplicateFunctionError, EntrySubsort, FunctionSymbol, LabSymbol, Scope,
    Symbol, SymbolCompareName, SymbolEntry,
};
use rugra::space::{space_flags, AddrSpace, SpaceType};
use rugra::type_system::datatype::TypeMetatype;
use rugra::type_system::typefactory::{CoreTypeFlavor, TypeFactory};

fn hexoff(v: u64) -> String {
    format!("{:x}", v)
}

fn main() {
    let mut types = TypeFactory::new_flavor(8, CoreTypeFlavor::Standalone);
    let int4 = types.get_base(4, TypeMetatype::Int).unwrap();
    let ram = AddrSpace::new_space(
        SpaceType::Processor,
        "ram",
        false,
        4,
        1,
        3,
        space_flags::HASPHYSICAL,
        -1,
        -1,
    );
    let global_id = 100u64;

    // ---- FunctionSymbol (database.cc:514/534/545/508) ----
    {
        let f = FunctionSymbol::new(global_id, "main", 2, Address::new(0));
        let fs = &f.symbol;
        println!(
            "case=fsym_buildtype|metatype=code|size={}|namelock={}|typelock={}|name={}",
            fs.dtype.as_ref().unwrap().get_size(),
            (fs.flags & symbol_flags::NAMELOCK != 0) as u8,
            (fs.flags & symbol_flags::TYPELOCK != 0) as u8,
            fs.name
        );
        println!("case=fsym_bytes_consumed|{}", f.get_bytes_consumed());
        println!("case=fsym_type_size|{}", fs.dtype.as_ref().unwrap().get_size());
        // The C++ decode ctor FunctionSymbol(Scope*,int4) (database.cc:545)
        // observable state, reached through the named ctor with an empty
        // name in the value model (identical fields).
        let fd = FunctionSymbol::new(global_id, "", 3, Address::new(0));
        println!(
            "case=fsym_decode_ctor|name_empty={}|consume={}",
            (fd.symbol.name.is_empty()) as u8,
            fd.get_bytes_consumed()
        );
    }

    // ---- LabSymbol (database.cc:728/736/745) ----
    {
        let l = LabSymbol::new(global_id, "loop", Address::new(0), &types);
        let dt = l.symbol.dtype.as_ref().unwrap();
        println!(
            "case=lsym_buildtype|typename={}|size={}|name={}",
            dt.get_name(),
            dt.get_size(),
            l.symbol.name
        );
        let ld = LabSymbol::new_decode(global_id, Address::new(0), &types);
        let dt = ld.symbol.dtype.as_ref().unwrap();
        println!(
            "case=lsym_decode_ctor|name_empty={}|typename={}",
            (ld.symbol.name.is_empty()) as u8,
            dt.get_name()
        );
    }

    // ---- ExternRefSymbol (database.cc:768/789; hh:351) ----
    {
        let x = rugra::database::ExternRefSymbol::new(global_id, "", Address::with_space(&ram, 0x6000));
        println!(
            "case=exref_autoname|name={}|externref={}|typelock={}|ptr_size={}|ptr_meta=ptr",
            x.symbol.name,
            (x.symbol.flags & symbol_flags::EXTERNREF != 0) as u8,
            (x.symbol.flags & symbol_flags::TYPELOCK != 0) as u8,
            x.symbol.dtype.as_ref().unwrap().get_size()
        );
        let y = rugra::database::ExternRefSymbol::new(global_id, "printf", Address::with_space(&ram, 0x6010));
        println!("case=exref_named|name={}", y.symbol.name);
        println!("case=exref_ref_addr|{}", hexoff(x.get_ref_addr().as_u64()));
    }

    // ---- UnionFacetSymbol (database.cc:691; hh:323/324) ----
    {
        let udt = types.get_type_union("fixture_union");
        let u = rugra::database::UnionFacetSymbol::new(global_id, "f", Some(udt), 2);
        println!(
            "case=ufacet_ctor|field={}|category={}|typename={}",
            u.get_field_number(),
            u.symbol.category as i32,
            u.symbol.dtype.as_ref().unwrap().get_name()
        );
        let ud = rugra::database::UnionFacetSymbol::new_decode(global_id);
        println!(
            "case=ufacet_decode|field={}|category={}",
            ud.get_field_number(),
            ud.symbol.category as i32
        );
    }

    // ---- Symbol::getBytesConsumed (database.cc:508) ----
    {
        let mut s = Symbol::new(global_id, "s", "int");
        s.dtype = Some(int4.clone());
        println!("case=sym_bytes_consumed|{}", s.get_bytes_consumed());
    }

    // ---- Symbol::getMapEntryPosition (database.cc:301-313) ----
    {
        let sym = std::sync::Arc::new(std::sync::RwLock::new(Symbol::new(global_id, "m", "int")));
        sym.write().unwrap().dtype = Some(int4.clone());
        let empty = RangeList::new();
        // Same insertion order as the C++ mapentry list: three whole
        // entries then a partial piece.
        let e1 = SymbolEntry::new_static(sym.clone(), 0, Address::with_space(&ram, 0x1000), 0, 4, empty.clone());
        let e2 = SymbolEntry::new_static(sym.clone(), 0, Address::with_space(&ram, 0x2000), 0, 4, empty.clone());
        let e3 = SymbolEntry::new_static(sym.clone(), 0, Address::with_space(&ram, 0x3000), 0, 4, empty.clone());
        let piece = SymbolEntry::new_static(sym.clone(), 0, Address::with_space(&ram, 0x4000), 0, 2, empty.clone());
        let entries = vec![e1, e2, e3, piece];
        let sym_r = sym.read().unwrap();
        println!("case=mapentry_whole_sought|{}", sym_r.get_map_entry_position(&entries, &entries[1]));
        println!("case=mapentry_partial_sought|{}", sym_r.get_map_entry_position(&entries, &entries[3]));
        // An entry of a DIFFERENT symbol is absent from the list.
        let other = std::sync::Arc::new(std::sync::RwLock::new(Symbol::new(global_id, "other", "int")));
        other.write().unwrap().dtype = Some(int4.clone());
        let foreign = SymbolEntry::new_static(other, 0, Address::with_space(&ram, 0x8000), 0, 4, empty);
        println!("case=mapentry_absent|{}", sym_r.get_map_entry_position(&entries, &foreign));
    }

    // ---- SymbolEntry::getFirstUseAddress (database.cc:122) ----
    {
        let sym = std::sync::Arc::new(std::sync::RwLock::new(Symbol::new(global_id, "m2", "int")));
        sym.write().unwrap().dtype = Some(int4.clone());
        let mut uselim = RangeList::new();
        uselim.insert_range(
            Range::new(Address::with_space(&ram, 0x4000), Address::with_space(&ram, 0x4fff))
                .unwrap(),
        );
        let e = SymbolEntry::new_static(sym.clone(), 0, Address::with_space(&ram, 0x2000), 0, 4, uselim);
        println!("case=firstuse_present|{}", hexoff(e.get_first_use_address().as_u64()));
        let e2 = SymbolEntry::new_static(sym, 0, Address::with_space(&ram, 0x2100), 0, 4, RangeList::new());
        println!("case=firstuse_empty|{}", e2.get_first_use_address().is_invalid() as u8);
    }

    // ---- SymbolEntry::printEntry (database.cc:166-181) ----
    {
        let sym = std::sync::Arc::new(std::sync::RwLock::new(Symbol::new(global_id, "m", "int")));
        sym.write().unwrap().dtype = Some(int4.clone());
        let e1 = SymbolEntry::new_static(sym.clone(), 0, Address::with_space(&ram, 0x1000), 0, 4, RangeList::new());
        print!("case=printentry_static_all|{}", e1.print_entry());

        let sym2 = std::sync::Arc::new(std::sync::RwLock::new(Symbol::new(global_id, "m2", "int")));
        sym2.write().unwrap().dtype = Some(int4.clone());
        let mut uselim = RangeList::new();
        uselim.insert_range(
            Range::new(Address::with_space(&ram, 0x4000), Address::with_space(&ram, 0x4fff))
                .unwrap(),
        );
        let e2 = SymbolEntry::new_static(sym2, 0, Address::with_space(&ram, 0x2000), 0, 4, uselim);
        print!("case=printentry_uselimit|{}", e2.print_entry());

        let symd = std::sync::Arc::new(std::sync::RwLock::new(Symbol::new(global_id, "d", "int")));
        symd.write().unwrap().dtype = Some(int4.clone());
        let e3 = SymbolEntry::new_dynamic(symd, 0, 0x1234, 0, 4, RangeList::new());
        print!("case=printentry_dynamic|{}", e3.print_entry());
    }

    // ---- SymbolEntry::EntrySubsort (database.hh:107-134) ----
    {
        let earliest = EntrySubsort::earliest();
        let latest = EntrySubsort::from_bool(true);
        println!("case=subsort_earliest_lt_latest|{}", earliest.lt(&latest) as u8);
        println!("case=subsort_latest_gt_earliest|{}", (!latest.lt(&earliest)) as u8);
        let from_a = EntrySubsort::from_parts(3, 0x5000);
        let from_b = EntrySubsort::from_parts(3, 0x6000);
        println!("case=subsort_addr_offset_order|{}", from_a.lt(&from_b) as u8);
        println!("case=subsort_addr_offset_rev|{}", (!from_b.lt(&from_a)) as u8);
        let from_a2 = EntrySubsort::from_parts(3, 0x5000);
        println!("case=subsort_addr_equal|{}", (!from_a.lt(&from_a2)) as u8);
    }

    // ---- SymbolCompareName (database.hh:366) ----
    {
        let mut s1 = Symbol::new(global_id, "apple", "int");
        s1.dtype = Some(int4.clone());
        let mut s2 = Symbol::new(global_id, "banana", "int");
        s2.dtype = Some(int4.clone());
        let cmp = SymbolCompareName;
        println!("case=symcmp_name_order|{}", cmp.is_before(&s1, &s2) as u8);
        println!("case=symcmp_name_rev|{}", (!cmp.is_before(&s2, &s1)) as u8);
        let mut t1 = Symbol::new(global_id, "same", "int");
        t1.dtype = Some(int4.clone());
        t1.name_dedup = 1;
        let mut t2 = Symbol::new(global_id, "same", "int");
        t2.dtype = Some(int4.clone());
        t2.name_dedup = 2;
        println!("case=symcmp_dedup_tiebreak|{}", cmp.is_before(&t1, &t2) as u8);
        println!("case=symcmp_dedup_rev|{}", (!cmp.is_before(&t2, &t1)) as u8);
    }

    // ---- DuplicateFunctionError (database.hh:435) ----
    {
        let err = DuplicateFunctionError::new(Address::with_space(&ram, 0x5000), "foo");
        println!(
            "case=dupfn|msg={}|addr={}|name={}",
            err.message,
            hexoff(err.address.as_u64()),
            err.function_name
        );
    }

    // ---- Scope::printBounds (database.hh:789) ----
    {
        let mut scope = Scope::new(101, "ns", global_id);
        let mut rlist = RangeList::new();
        rlist.insert_range(
            Range::new(Address::with_space(&ram, 0x3000), Address::with_space(&ram, 0x3fff))
                .unwrap(),
        );
        rlist.insert_range(
            Range::new(Address::with_space(&ram, 0x1000), Address::with_space(&ram, 0x1fff))
                .unwrap(),
        );
        for rng in rlist.ranges() {
            scope.rangetree.insert_range(*rng);
        }
        print!("case=scope_printbounds|{}", scope.print_bounds());
    }
}
